src/continuous_agg.c

```diff
@@ -935,13 +1009,12 @@ ts_continuous_agg_watermark(PG_FUNCTION_ARGS)
 
 	if (watermark != NULL)
 	{
-		/* The flinfo is probably reset when the input argument changes, but
-		 * the documentation isn't clear on this. Check that the hypertable ID
-		 * matches across repeated calls just to be safe. */
-		if (watermark->hyper_id == hyper_id)
+		/* Check that the hypertable ID matches across repeated calls. Reset
+		 * the cached watermark if the ID changes. */
+		if (watermark_valid(watermark, hyper_id))
 			PG_RETURN_INT64(watermark->value);
 
-		pfree(watermark);
+		MemoryContextDelete(watermark->mctx);
```

Comment:
This one will also be called when `top_mctx` gets deleted as well, right?

Reply:
Yes. If you destroy a parent, then all its children will be deleted.